File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/codesigning/datastructure/src/code_sign_block.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "code_sign_block.h"
#include "merkle_tree_extension.h"

namespace OHOS {
namespace SignatureTools {

CodeSignBlock::CodeSignBlock()
{
}

CodeSignBlock::~CodeSignBlock()
{
}

void CodeSignBlock::AddOneMerkleTree(const std::string& key, const std::vector<int8_t>& merkleTree)
{
    if (key.empty()) {
        return;
    }
    if (merkleTreeMap.find(key) == merkleTreeMap.end()) {
        if (merkleTree.empty()) {
            merkleTreeMap.insert(std::make_pair(key, std::vector<int8_t>(0)));
        } else {
            merkleTreeMap.insert(std::make_pair(key, merkleTree));
        }
    }
}

std::vector<int8_t> CodeSignBlock::GetOneMerkleTreeByFileName(const std::string& key)
{
    if (key.empty()) {
        return std::vector<int8_t>();
    }
    return merkleTreeMap[key];
}

void CodeSignBlock::SetCodeSignBlockFlag()
{
    int flags = CodeSignBlockHeader::FLAG_MERKLE_TREE_INLINED;
    if (nativeLibInfoSegment.GetSectionNum() != 0) {
        flags |= CodeSignBlockHeader::FLAG_NATIVE_LIB_INCLUDED;
    }
    codeSignBlockHeader.SetFlags(flags);
}

void CodeSignBlock::SetSegmentNum()
{
    codeSignBlockHeader.SetSegmentNum(static_cast<int>(segmentHeaderList.size()));
}

void CodeSignBlock::AddToSegmentList(const SegmentHeader& sh)
{
    segmentHeaderList.push_back(sh);
}

std::vector<SegmentHeader>& CodeSignBlock::GetSegmentHeaderList()
{
    return segmentHeaderList;
}

void CodeSignBlock::SetSegmentHeaders()
{
    // fs-verity info segment
    SegmentHeader tempVar(SegmentHeader::CSB_FSVERITY_INFO_SEG, FsVerityInfoSegment::FS_VERITY_INFO_SEGMENT_SIZE);
    segmentHeaderList.push_back(tempVar);
    // hap info segment
    SegmentHeader tempVar2(SegmentHeader::CSB_HAP_META_SEG, hapInfoSegment.GetSize());
    segmentHeaderList.push_back(tempVar2);
    // native lib info segment
    SegmentHeader tempVar3(SegmentHeader::CSB_NATIVE_LIB_INFO_SEG, nativeLibInfoSegment.Size());
    segmentHeaderList.push_back(tempVar3);
}

CodeSignBlockHeader& CodeSignBlock::GetCodeSignBlockHeader()
{
    return codeSignBlockHeader;
}

void CodeSignBlock::SetCodeSignBlockHeader(const CodeSignBlockHeader& csbHeader)
{
    codeSignBlockHeader = csbHeader;
}

void CodeSignBlock::SetFsVerityInfoSegment(const FsVerityInfoSegment& fsVeritySeg)
{
    fsVerityInfoSegment = fsVeritySeg;
}

HapInfoSegment& CodeSignBlock::GetHapInfoSegment()
{
    return hapInfoSegment;
}

void CodeSignBlock::SetHapInfoSegment(const HapInfoSegment& hapSeg)
{
    hapInfoSegment = hapSeg;
}

NativeLibInfoSegment& CodeSignBlock::GetSoInfoSegment()
{
    return nativeLibInfoSegment;
}

void CodeSignBlock::SetSoInfoSegment(const NativeLibInfoSegment& soSeg)
{
    nativeLibInfoSegment = soSeg;
}

void CodeSignBlock::ToByteArray(std::vector<int8_t>& ret)
{
    std::unique_ptr<ByteBuffer> bf = std::make_unique<ByteBuffer>
        (ByteBuffer(codeSignBlockHeader.GetBlockSize()));
    std::vector<int8_t> codeSignBlockHeaderVec;
    codeSignBlockHeader.ToByteArray(codeSignBlockHeaderVec);
    bf->PutData(codeSignBlockHeaderVec.data(), codeSignBlockHeaderVec.size());
    for (auto &sh : segmentHeaderList) {
        std::vector<int8_t> shVec;
        sh.ToByteArray(shVec);
        bf->PutData(shVec.data(), shVec.size());
    }
    bf->PutData(zeroPadding.data(), zeroPadding.size());
    // Hap merkle tree
    bf->PutData(merkleTreeMap["Hap"].data(), merkleTreeMap["Hap"].size());
    std::vector<int8_t> fsVerityInfoSegmentVec;
    fsVerityInfoSegment.ToByteArray(fsVerityInfoSegmentVec);
    bf->PutData(fsVerityInfoSegmentVec.data(), fsVerityInfoSegmentVec.size());
    std::vector<int8_t> hapInfoSegmentVec;
    hapInfoSegment.ToByteArray(hapInfoSegmentVec);
    bf->PutData(hapInfoSegmentVec.data(), hapInfoSegmentVec.size());
    std::vector<int8_t> nativeLibInfoSegmentVec;
    nativeLibInfoSegment.ToByteArray(nativeLibInfoSegmentVec);
    bf->PutData(nativeLibInfoSegmentVec.data(), nativeLibInfoSegmentVec.size());
    
    ret = std::vector<int8_t>(bf->GetBufferPtr(), bf->GetBufferPtr() + bf->GetPosition());
}

void CodeSignBlock::ComputeSegmentOffset()
{
    // 1) the first segment is placed after merkle tree
    int segmentOffset = CodeSignBlockHeader::Size()
        + segmentHeaderList.size() * SegmentHeader::SEGMENT_HEADER_LENGTH
        + zeroPadding.size() + GetOneMerkleTreeByFileName("Hap").size();
    for (std::vector<SegmentHeader>::size_type i = 0; i < segmentHeaderList.size(); i++) {
        segmentHeaderList[i].SetSegmentOffset(static_cast<int32_t>(segmentOffset));
        segmentOffset += segmentHeaderList[i].GetSegmentSize();
    }
}

int64_t CodeSignBlock::ComputeMerkleTreeOffset(int64_t codeSignBlockOffset)
{
    int64_t sizeWithoutMerkleTree = CodeSignBlockHeader::Size()
        + SEGMENT_HEADER_COUNT * SegmentHeader::SEGMENT_HEADER_LENGTH;
    // add code sign block offset while computing align position for merkle tree
    int64_t residual = (codeSignBlockOffset + sizeWithoutMerkleTree) % PAGE_SIZE_4K;
    if (residual == 0) {
        zeroPadding = std::vector<int8_t>(0);
    } else {
        zeroPadding = std::vector<int8_t>(static_cast<int>(PAGE_SIZE_4K - residual));
    }
    return codeSignBlockOffset + sizeWithoutMerkleTree + zeroPadding.size();
}

void CodeSignBlock::GenerateCodeSignBlockByte(int64_t fsvTreeOffset, std::vector<int8_t>& ret)
{
    // 1) compute overall block size without merkle tree
    int64_t csbSize = CodeSignBlockHeader::Size()
        + static_cast<int64_t>(segmentHeaderList.size()) * SegmentHeader::SEGMENT_HEADER_LENGTH
        + zeroPadding.size()
        + GetOneMerkleTreeByFileName("Hap").size()
        + fsVerityInfoSegment.Size()
        + hapInfoSegment.GetSize()
        + nativeLibInfoSegment.Size();
    Extension* ext = hapInfoSegment.GetSignInfo().GetExtensionByType(MerkleTreeExtension::MERKLE_TREE_INLINED);
    if (ext != nullptr) {
        MerkleTreeExtension* merkleTreeExtension = static_cast<MerkleTreeExtension*>(ext);
        merkleTreeExtension->SetMerkleTreeOffset(fsvTreeOffset);
    }
    codeSignBlockHeader.SetBlockSize(csbSize);
    // 2) generate byte array of complete code sign block
    ToByteArray(ret);
}

}
}

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/codesigning/datastructure/CodeSignBlock.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.codesigning.datastructure;

import com.ohos.hapsigntool.codesigning.sign.CodeSigning;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * code sign block is a chunk of bytes attached to hap package or file.
 * It consists of two headers:
 * 1) code sign block header
 * 2) segment header
 * three segments:
 * 1) fs-verity info segment
 * 2) hap info segment
 * 3) so info segment
 * one zero padding area in order to align merkle tree raw bytes
 * 1) zero padding
 * and one area storing merkle tree bytes:
 * 1) merkle tree raw bytes
 * <p>
 * After signing a hap, call toByteArray() method to generate a block of bytes.
 *
 * @since 2023/09/08
 */
public class CodeSignBlock {
    /**
     * page size in bytes
     */
    public static final long PAGE_SIZE_4K = 4096L;

    /**
     * Segment header count, including fs-verity info, hap info, so info segment
     */
    public static final int SEGMENT_HEADER_COUNT = 3;

    private CodeSignBlockHeader codeSignBlockHeader;

    private final List<SegmentHeader> segmentHeaderList;

    private FsVerityInfoSegment fsVerityInfoSegment;

    private HapInfoSegment hapInfoSegment;

    private NativeLibInfoSegment nativeLibInfoSegment;

    private byte[] zeroPadding;

    private final Map<String, byte[]> merkleTreeMap;

    public CodeSignBlock() {
        this.codeSignBlockHeader = new CodeSignBlockHeader.Builder().build();
        this.segmentHeaderList = new ArrayList<>();
        this.fsVerityInfoSegment = new FsVerityInfoSegment();
        this.hapInfoSegment = new HapInfoSegment();
        this.nativeLibInfoSegment = new NativeLibInfoSegment.Builder().build();
        this.merkleTreeMap = new HashMap<>();
    }

    /**
     * Add one merkle tree into merkleTreeMap
     *
     * @param key        file name
     * @param merkleTree merkle tree raw bytes
     */
    public void addOneMerkleTree(String key, byte[] merkleTree) {
        if (merkleTree == null) {
            this.merkleTreeMap.put(key, new byte[0]);
        } else {
            this.merkleTreeMap.put(key, merkleTree);
        }
    }

    /**
     * Get one merkle tree from merkleTreeMap by file name
     *
     * @param key file name
     * @return merkle tree bytes
     */
    public byte[] getOneMerkleTreeByFileName(String key) {
        return this.merkleTreeMap.get(key);
    }

    /**
     * set code sign block flag
     */
    public void setCodeSignBlockFlag() {
        int flags = CodeSignBlockHeader.FLAG_MERKLE_TREE_INLINED;
        if (this.nativeLibInfoSegment.getSectionNum() != 0) {
            flags += CodeSignBlockHeader.FLAG_NATIVE_LIB_INCLUDED;
        }
        this.codeSignBlockHeader.setFlags(flags);
    }

    /**
     * set segmentNum defined in code sign block header, equals length of segmentHeaderList
     */
    public void setSegmentNum() {
        this.codeSignBlockHeader.setSegmentNum(segmentHeaderList.size());
    }

    /**
     * add one segment to segmentHeaderList
     *
     * @param sh segment header
     */
    public void addToSegmentList(SegmentHeader sh) {
        this.segmentHeaderList.add(sh);
    }

    public List<SegmentHeader> getSegmentHeaderList() {
        return segmentHeaderList;
    }

    /**
     * set segment header list
     */
    public void setSegmentHeaders() {
        // fs-verity info segment
        segmentHeaderList.add(new SegmentHeader(SegmentHeader.CSB_FSVERITY_INFO_SEG, this.fsVerityInfoSegment.size()));
        // hap info segment
        segmentHeaderList.add(new SegmentHeader(SegmentHeader.CSB_HAP_META_SEG, this.hapInfoSegment.size()));
        // native lib info segment
        segmentHeaderList.add(
            new SegmentHeader(SegmentHeader.CSB_NATIVE_LIB_INFO_SEG, this.nativeLibInfoSegment.size()));
    }

    public CodeSignBlockHeader getCodeSignBlockHeader() {
        return codeSignBlockHeader;
    }

    public void setCodeSignBlockHeader(CodeSignBlockHeader csbHeader) {
        this.codeSignBlockHeader = csbHeader;
    }

    public void setFsVerityInfoSegment(FsVerityInfoSegment fsVeritySeg) {
        this.fsVerityInfoSegment = fsVeritySeg;
    }

    public FsVerityInfoSegment getFsVerityInfoSegment() {
        return fsVerityInfoSegment;
    }

    public HapInfoSegment getHapInfoSegment() {
        return hapInfoSegment;
    }

    public void setHapInfoSegment(HapInfoSegment hapSeg) {
        this.hapInfoSegment = hapSeg;
    }

    public NativeLibInfoSegment getSoInfoSegment() {
        return nativeLibInfoSegment;
    }

    public void setSoInfoSegment(NativeLibInfoSegment soSeg) {
        this.nativeLibInfoSegment = soSeg;
    }

    /**
     * Convert code sign block object to a newly created byte array
     *
     * @return Byte array representation of a CodeSignBlock object
     */
    public byte[] toByteArray() {
        ByteBuffer bf = ByteBuffer.allocate(this.codeSignBlockHeader.getBlockSize()).order(ByteOrder.LITTLE_ENDIAN);
        bf.put(this.codeSignBlockHeader.toByteArray());
        for (SegmentHeader sh : this.segmentHeaderList) {
            bf.put(sh.toByteArray());
        }
        bf.put(this.zeroPadding);
        // Hap merkle tree
        if (this.hapInfoSegment.getSignInfo().getExtensionByType(MerkleTreeExtension.MERKLE_TREE_INLINED) != null) {
            bf.put(merkleTreeMap.get("Hap"));
        }
        bf.put(this.fsVerityInfoSegment.toByteArray());
        bf.put(this.hapInfoSegment.toByteArray());
        bf.put(this.nativeLibInfoSegment.toByteArray());
        return bf.array();
    }

    /**
     * SegmentOffset is the position of each segment defined in segmentHeaderList,
     * based on the start position of code sign block
     */
    public void computeSegmentOffset() {
        // 1) the first segment is placed after merkle tree
        int segmentOffset = CodeSignBlockHeader.size()
            + this.segmentHeaderList.size() * SegmentHeader.SEGMENT_HEADER_LENGTH + this.zeroPadding.length
            + this.getOneMerkleTreeByFileName(CodeSigning.HAP_SIGNATURE_ENTRY_NAME).length;
        for (SegmentHeader sh : segmentHeaderList) {
            sh.setSegmentOffset(segmentOffset);
            segmentOffset += sh.getSegmentSize();
        }
    }

    /**
     * Compute the offset to store merkle tree raw bytes based on file start
     *
     * @param codeSignBlockOffset offset to store code sign block based on file start
     * @return offset to store merkle tree based on the file start, it includes the codeSignBlockOffset
     */
    public long computeMerkleTreeOffset(long codeSignBlockOffset) {
        long sizeWithoutMerkleTree = CodeSignBlockHeader.size()
            + SEGMENT_HEADER_COUNT * SegmentHeader.SEGMENT_HEADER_LENGTH;
        // add code sign block offset while computing align position for merkle tree
        long residual = (codeSignBlockOffset + sizeWithoutMerkleTree) % PAGE_SIZE_4K;
        if (residual == 0) {
            this.zeroPadding = new byte[0];
        } else {
            this.zeroPadding = new byte[(int) (PAGE_SIZE_4K - residual)];
        }
        return codeSignBlockOffset + sizeWithoutMerkleTree + zeroPadding.length;
    }

    /**
     * Convert CodeSignBlock to bytes
     *
     * @param fsvTreeOffset merkle tree offset
     * @return byte array representing the code sign block
     */
    public byte[] generateCodeSignBlockByte(long fsvTreeOffset) {
        // 1) compute overall block size without merkle tree
        long csbSize = CodeSignBlockHeader.size()
            + (long) this.segmentHeaderList.size() * SegmentHeader.SEGMENT_HEADER_LENGTH + this.zeroPadding.length
            + this.getOneMerkleTreeByFileName(CodeSigning.HAP_SIGNATURE_ENTRY_NAME).length
            + this.fsVerityInfoSegment.size() + this.hapInfoSegment.size() + this.nativeLibInfoSegment.size();
        Extension ext = this.hapInfoSegment.getSignInfo().getExtensionByType(MerkleTreeExtension.MERKLE_TREE_INLINED);
        if (ext instanceof MerkleTreeExtension) {
            MerkleTreeExtension merkleTreeExtension = (MerkleTreeExtension) ext;
            merkleTreeExtension.setMerkleTreeOffset(fsvTreeOffset);
        }
        this.codeSignBlockHeader.setBlockSize(csbSize);
        // 2) generate byte array of complete code sign block
        return toByteArray();
    }

    /**
     * Return a string representation of the object
     *
     * @return string representation of the object
     */
    public String toString() {
        return String.format(Locale.ROOT,
            "CodeSignBlockHeader[%s], SegmentHeaders[%s], FsVeritySeg[%s], HapInfoSeg[%s], SoInfoSeg[%s]",
            this.codeSignBlockHeader, Arrays.toString(this.segmentHeaderList.toArray()), this.fsVerityInfoSegment,
            this.hapInfoSegment, this.nativeLibInfoSegment);
    }
}
